presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultiJoinNode.java

```diff
@@ -0,0 +1,141 @@
+        this.filter = requireNonNull(filter, "filter is null");
+        this.outputSymbols = ImmutableList.copyOf(requireNonNull(outputSymbols, "outputSymbols is null"));
+
+        List<Symbol> inputSymbols = sources.stream().flatMap(source -> source.getOutputSymbols().stream()).collect(toImmutableList());
+        checkArgument(inputSymbols.containsAll(outputSymbols), "inputs do not contain all output symbols");
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(
+                ImmutableSet.copyOf(this.getOutputSymbols()),
+                ImmutableSet.copyOf(extractConjuncts(this.getFilter())),
+                this.getSources().stream()
+                        .map(PlanNode::getId)
+                        .collect(toImmutableSet()));
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof MultiJoinNode)) {
+            return false;
+        }
+
+        MultiJoinNode otherNode = (MultiJoinNode) other;
+        return ImmutableSet.copyOf(this.getOutputSymbols()).equals(ImmutableSet.copyOf(otherNode.getOutputSymbols()))
+                && ImmutableSet.copyOf(extractConjuncts(this.getFilter())).equals(ImmutableSet.copyOf(extractConjuncts(otherNode.getFilter())))
+                && this.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()).equals(otherNode.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()));
+    }
+
+    static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup)
+    {
+        return new MultiJoinNodeBuilder(joinNode, lookup).toMultiJoinNode();
+    }
+
+    private static class MultiJoinNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+        private final List<Symbol> outputSymbols;
+        private final Lookup lookup;
+
+        MultiJoinNodeBuilder(JoinNode node, Lookup lookup)
+        {
+            requireNonNull(node, "node is null");
+            checkState(node.getType() == INNER, "join type must be INNER");
+            this.outputSymbols = node.getOutputSymbols();
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            flattenNode(node);
+        }
+
+        private void flattenNode(PlanNode node)
+        {
+            PlanNode resolved = lookup.resolve(node);
+            if (resolved instanceof JoinNode && ((JoinNode) resolved).getType() == INNER && isDeterministic(((JoinNode) resolved).getFilter().orElse(TRUE_LITERAL))) {
+                JoinNode joinNode = (JoinNode) resolved;
+                flattenNode(joinNode.getLeft());
```

Comment:
> you can't prioritize rule execution within iterative optimizer

that's a valid point

> The MultiJoinNode would need to be a proper PlanNode even though it's very transient 

the `ApplyNode` is quite similar in this respect - it's a temporary, logical node that has no physical implementation other than a rewrite.

>with the limitation of having dedicated iteratitve optimizers just for that, I don't see benefits.

I think we can leave it as is then, as long as we shield the rules used in exploratory optimizer from lookup and memo using patterns.